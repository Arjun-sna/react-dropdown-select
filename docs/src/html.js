import React from "react"
import PropTypes from "prop-types"

export default class HTML extends React.Component {
  render() {
    return (
      <html {...this.props.htmlAttributes}>
        <head>
          <meta charSet="utf-8" />
          <meta httpEquiv="x-ua-compatible" content="ie=edge" />
          <meta
            name="viewport"
            content="width=device-width, initial-scale=1, shrink-to-fit=no"
          />
          {this.props.headComponents}
        </head>
        <body {...this.props.bodyAttributes}>
          {this.props.preBodyComponents}
          <div
            key={`body`}
            id="___gatsby"
            dangerouslySetInnerHTML={{ __html: this.props.body }}
          />
          {this.props.postBodyComponents}
            <link rel="stylesheet" href="https://d2v52k3cl9vedd.cloudfront.net/basscss/7.0.4/basscss.min.css"/>
            <link rel="stylesheet"
                  href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.0.0/styles/monokai.min.css"/>
        </body>
      </html>
    )
  }
}

HTML.propTypes = {
  htmlAttributes: PropTypes.object,
  headComponents: PropTypes.array,
  bodyAttributes: PropTypes.object,
  preBodyComponents: PropTypes.array,
  body: PropTypes.string,
  postBodyComponents: PropTypes.array,
}
