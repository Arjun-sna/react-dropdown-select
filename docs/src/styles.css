body, html {
	font-family: Arial, "Lucida Grande", sans-serif !important;
	font-weight: 300;
}

h1, h2, h3, h4, h5, h6 {
	font-family: Arial, "Lucida Grande", sans-serif;
	font-weight: 300;
}

.hp .material-icons {
	font-size: 70px;
}

blockquote {
	background: #fff;
	margin: 21px 0;
	border-left: 3px solid #ccc;
	padding: 0.5em 10px;
	quotes: "\201C" "\201D" "\2018" "\2019";
}

blockquote:before {
	color: #ccc;
	content: open-quote;
	font-size: 4em;
	line-height: 0.1em;
	margin-right: 0.25em;
	vertical-align: -0.4em;
}

blockquote p {
	display: inline;
}

.home-docs-link {
	width: 30%;
	margin: 0 auto;
	display: block;
	text-align: center;
}

table {
	font-family: Arial, sans-serif;
	color: #999;
}

table tr td:not(:nth-child(1)) {
	font-size: small;
}

th, td {
	border-bottom: 1px solid #ddd;
}


th, tr:hover {
	background: #f2f2f2;
	color: #000;
}

tr > td:nth-child(1) {
	font-weight: 500;
}

tr > td {
	font-family: "Fira Code", "Andale Mono", monospace;
}

summary {
	font-size: larger;
	margin: 20px 0;
}

pre, code {
	background: #f2f2f2;
	padding: 0 20px 20px;
	font-family: "Fira Code", "Andale Mono", monospace;
}
