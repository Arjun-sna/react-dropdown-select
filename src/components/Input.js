import React from 'react';
import styled from '@emotion/styled';

const handlePlaceHolder = (parentProps, parentState) => {
  const noValues = parentState.values && parentState.values.length === 0;
  const hasValues = parentState.values && parentState.values.length > 0;

  if (hasValues && parentProps.addPlaceholder && parentProps.searchable) {
    return parentProps.addPlaceholder;
  }

  if (noValues) {
    return parentProps.placeholder;
  }

  if (hasValues && !parentProps.searchable) {
    return '';
  }

  return '';
};

const Input = ({ parentProps, parentState, parentMethods }) => {
  return parentProps.inputRenderer ? (
    parentProps.inputRenderer(parentProps, parentState, parentMethods)
  ) : (
    <InputComponent
      tabIndex="-1"
      className="react-dropdown-select-input"
      size={parentMethods.getInputSize()}
      value={parentState.search}
      readOnly={!parentProps.searchable}
      onClick={() => parentMethods.dropDown('open')}
      onChange={parentMethods.setSearch}
      placeholder={handlePlaceHolder(parentProps, parentState)}
    />
  );
};

Input.propTypes = {};

const InputComponent = styled.input`
  flex: 1;
  display: flex;
  line-height: inherit;
  width: auto;
  border: none;
  margin-left: 5px;
  background: transparent;
  font-size: smaller;
  ${({ readOnly}) => readOnly && 'cursor: pointer;'}
  :focus {
    outline: none;
  }
`;

export default Input;
