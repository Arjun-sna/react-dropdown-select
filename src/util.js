export const valueExistInSelected = (value, values) => !!values.find((val) => val.value === value);

export const hexToRGBA = (hex, alpha) => {
  const RR = parseInt(hex.slice(1, 3), 16);
  const GG = parseInt(hex.slice(3, 5), 16);
  const BB = parseInt(hex.slice(5, 7), 16);

  return `rgba(${RR}, ${GG}, ${BB}${alpha && `, ${alpha}`})`;
};

export const debounce = (fn, delay = 0) => {
  let timerId;

  return (...args) => {
    if (timerId) {
      clearTimeout(timerId);
    }
    timerId = setTimeout(() => {
      fn(...args);
      timerId = null;
    }, delay);
  };
};
