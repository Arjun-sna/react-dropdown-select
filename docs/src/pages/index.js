import React from 'react';
import { Link } from 'gatsby';
import Header from '../components/header';
import Footer from '../components/footer';
import Select from '../../../src/index';
import styled from 'styled-components';

import '../styles.css';

export class Home extends React.Component {
  state = {
    multi: true,
    disabled: false,
    loading: false,
    contentRenderer: false,
    dropdownRenderer: false,
    inputRenderer: false,
    itemRenderer: false,
    optionRenderer: false,
    noDataRenderer: false,
    selectValues: [],
    searchBy: 'label',
    clearable: true,
    separator: true,
    forceOpen: false,
    handle: true,
    options: []
  };

  componentDidMount() {
    window
      .fetch('https://jsonplaceholder.typicode.com/users')
      .then((responce) => responce.json())
      .then((result) => {
        this.setOptions(
          result.map((user) => ({
            label: user.username,
            value: user.email
          }))
        );
      });
  }

  setValues = (selectValues) => this.setState({ selectValues });

  setOptions = (options) => this.setState({ options });

  contentRenderer = (innerProps, innerState) => {
    return (
      <div>
        {innerState.values.length} of {innerProps.options.length} Selected
      </div>
    );
  };

  noDataRenderer = () => {
    return (
      <p style={{ textAlign: 'center' }}>
        <strong>Ooops!</strong> No data found
      </p>
    );
  };

  itemRenderer = (item, itemIndex, props, state, methods) => (
    <div key={item.value}>
      <div style={{ margin: '10px' }}>
        <input
          onChange={() => methods.addItem(item)}
          type="checkbox"
          checked={methods.isSelected(item)}
        />
        <span onClick={() => methods.addItem(item)}>&nbsp;&nbsp;&nbsp;{item.label}</span>
      </div>
    </div>
  );

  dropdownRenderer = (props, state, methods) => {
    const regexp = new RegExp(state.search, 'i');

    return (
      <div>
        <input
          type="text"
          size={methods.getInputSize()}
          value={state.search}
          onChange={methods.setSearch}
          placeholder="Type anything"
        />
        <button onClick={methods.selectAll}>Select all</button>
        <button onClick={methods.clearAll}>Clear all</button>
        {props.options
          .filter((item) => regexp.test(item[props.searchBy] || item.label))
          .map((option) => (
            <div key={option.value} onClick={() => methods.addItem(option)}>
              <input
                type="checkbox"
                onChange={() => methods.addItem(option)}
                checked={state.values.indexOf(option) !== -1}
              />
              {option.label}
            </div>
          ))}
      </div>
    );
  };

  optionRenderer = (option, props, state, methods) => (
    <React.Fragment>
      <div onClick={(event) => methods.removeItem(event, option, true)}>{option.label}</div>
    </React.Fragment>
  );

  inputRenderer = (props, state, methods) => (
    <input
      tabIndex="1"
      className="react-dropdown-select-input"
      size={methods.getInputSize()}
      value={state.search}
      onClick={() => methods.dropDown('open')}
      onChange={methods.setSearch}
      placeholder="Type in"
    />
  );

  render() {
    return (
      <div className="container px2 sm-px3">
        <Header page="home" />

        <article className="markdown-body px0">
          <h1 id="what-it-is-all-about">Customisable dropdown select for react</h1>

          <p>
            Customisable dropdown select for react with custom render callback props to override
            inner components
          </p>

          <h3>
            Install via <strong>npm</strong>:
          </h3>

          <pre>npm install --save react-dropdown-select</pre>

          <h3>Use:</h3>
          <pre>
            {`
<Select
  multi={true}
  options={options}
  onChange={(values) => this.onChange(values)}
/>`}
          </pre>

          <p className="center">
            <Link to="props" className="m1 btn btn-outline center">
              Prop types
            </Link>
          </p>

          <StyledWrapper>
            <Select
              placeholder="Select user"
              addPlaceholder="+ click to add"
              disabled={this.state.disabled}
              loading={!this.state.options.length || this.state.loading}
              searchBy={this.state.searchBy}
              separator={this.state.separator}
              clearable={this.state.clearable}
              keepOpen={this.state.forceOpen}
              dropdownHandle={this.state.handle}
              multi={this.state.multi}
              values={[]}
              options={this.state.options}
              dropdownGap={5}
              onDropdownOpen={() => undefined}
              onDropdownClose={() => undefined}
              onClearAll={() => undefined}
              onSelectAll={() => undefined}
              onChange={(values) => this.setValues(values)}
              noDataLabel="No matches found"
              noDataRenderer={this.state.noDataRenderer ? () => this.noDataRenderer() : undefined}
              itemRenderer={
                this.state.itemRenderer
                  ? (item, itemIndex, props, state, methods) =>
                      this.itemRenderer(item, itemIndex, props, state, methods)
                  : undefined
              }
              inputRenderer={
                this.state.inputRenderer
                  ? (props, state, methods) => this.inputRenderer(props, state, methods)
                  : undefined
              }
              optionRenderer={
                this.state.optionRenderer
                  ? (option, props, state, methods) =>
                      this.optionRenderer(option, props, state, methods)
                  : undefined
              }
              contentRenderer={
                this.state.contentRenderer
                  ? (innerProps, innerState) => this.contentRenderer(innerProps, innerState)
                  : undefined
              }
              dropdownRenderer={
                this.state.dropdownRenderer
                  ? (innerProps, innerState, innerMethods) =>
                      this.dropdownRenderer(innerProps, innerState, innerMethods)
                  : undefined
              }
            />
          </StyledWrapper>

          <p>
            <input
              type="checkbox"
              checked={this.state.multi}
              onChange={() =>
                this.setState({
                  multi: !this.state.multi
                })
              }
            />{' '}
            Multi
            <br />
            <input
              type="checkbox"
              checked={this.state.disabled}
              onChange={() =>
                this.setState({
                  disabled: !this.state.disabled
                })
              }
            />{' '}
            Disabled
            <br />
            <input
              type="checkbox"
              checked={this.state.loading}
              onChange={() =>
                this.setState({
                  loading: !this.state.loading
                })
              }
            />{' '}
            Loading
            <br />
            <input
              type="checkbox"
              checked={this.state.clearable}
              onChange={() =>
                this.setState({
                  clearable: !this.state.clearable
                })
              }
            />{' '}
            Clearable
            <br />
            <input
              type="checkbox"
              checked={this.state.separator}
              onChange={() =>
                this.setState({
                  separator: !this.state.separator
                })
              }
            />{' '}
            Separator
            <br />
            <input
              type="checkbox"
              checked={this.state.handle}
              onChange={() =>
                this.setState({
                  handle: !this.state.handle
                })
              }
            />{' '}
            Dropdown handle
            <br />
            <input
              type="checkbox"
              checked={this.state.forceOpen}
              onChange={() =>
                this.setState({
                  forceOpen: !this.state.forceOpen
                })
              }
            />{' '}
            Stay open
            <br />
            <input
              type="checkbox"
              checked={this.state.contentRenderer}
              onChange={() =>
                this.setState({
                  contentRenderer: !this.state.contentRenderer
                })
              }
            />{' '}
            Custom content renderer
            <br />
            <input
              type="checkbox"
              checked={this.state.dropdownRenderer}
              onChange={() =>
                this.setState({
                  dropdownRenderer: !this.state.dropdownRenderer
                })
              }
            />{' '}
            Custom dropdown renderer |{' '}
            <a href="https://codesandbox.io/s/5k0y843684">demo @ codesandbox</a>
            <br />
            <input
              type="checkbox"
              checked={this.state.itemRenderer}
              onChange={() =>
                this.setState({
                  itemRenderer: !this.state.itemRenderer
                })
              }
            />{' '}
            Custom dropdown item renderer
            <br />
            Search by field:{' '}
            <select
              selected={this.state.searchBy}
              onChange={(searchBy) =>
                this.setState({
                  searchBy
                })
              }>
              <option value="label">label</option>
              <option value="customProperty">customProperty</option>
            </select>
          </p>

          <p>Current value(s):</p>
          <pre>{JSON.stringify(this.state.selectValues, false, 2)}</pre>

          <p>
            Options <small>(loaded from: https://jsonplaceholder.typicode.com/users)</small>:
          </p>
          <pre>{JSON.stringify(this.state.options, false, 2)}</pre>
        </article>

        <Footer />
      </div>
    );
  }
}

export default Home;

const StyledWrapper = styled.div`
  width: 400px;
  margin: 20px auto;
`;
