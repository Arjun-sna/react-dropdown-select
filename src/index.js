import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import ClickOutHandler from 'react-onclickout';

import Content from './components/Content';
import Dropdown from './components/Dropdown';
import Loading from './components/Loading';
import Clear from './components/Clear';
import Separator from './components/Separator';
import DropdownHandle from './components/DropdownHandle';

import { debounce, hexToRGBA } from './util';

export class Select extends Component {
  static propTypes = {
    onChange: PropTypes.func.isRequired,
    onDropdownClose: PropTypes.func,
    onDropdownOpen: PropTypes.func,
    onClearAll: PropTypes.func,
    onSelectAll: PropTypes.func,
    values: PropTypes.array,
    options: PropTypes.array.isRequired,
    keepOpen: PropTypes.bool,
    dropdownGap: PropTypes.number,
    multi: PropTypes.bool,
    placeholder: PropTypes.string,
    addPlaceholder: PropTypes.string,
    disabled: PropTypes.bool,
    className: PropTypes.string,
    loading: PropTypes.bool,
    clearable: PropTypes.bool,
    searchable: PropTypes.bool,
    separator: PropTypes.bool,
    dropdownHandle: PropTypes.bool,
    searchBy: PropTypes.string,
    sortBy: PropTypes.string,
    closeOnScroll: PropTypes.bool,
    openOnTop: PropTypes.bool,
    style: PropTypes.object,
    contentRenderer: PropTypes.func,
    dropdownRenderer: PropTypes.func,
    itemRenderer: PropTypes.func,
    noDataRenderer: PropTypes.func,
    optionRenderer: PropTypes.func,
    inputRenderer: PropTypes.func,
    loadingRenderer: PropTypes.func,
    clearRenderer: PropTypes.func,
    separatorRenderer: PropTypes.func,
    dropdownHandleRenderer: PropTypes.func
  };

  constructor(props) {
    super(props);

    this.state = {
      dropdown: false,
      values: props.values,
      search: '',
      selectBounds: {},
      cursor: null
    };

    this.methods = {
      removeItem: this.removeItem,
      dropDown: this.dropDown,
      addItem: this.addItem,
      setSearch: this.setSearch,
      getInputSize: this.getInputSize,
      toggleSelectAll: this.toggleSelectAll,
      clearAll: this.clearAll,
      selectAll: this.selectAll,
      searchResults: this.searchResults,
      getSelectRef: this.getSelectRef,
      isSelected: this.isSelected,
      getSelectBounds: this.getSelectBounds,
      areAllSelected: this.areAllSelected,
      handleKeyDown: this.handleKeyDown,
      activeCursorItem: this.activeCursorItem,
      createNew: this.createNew
    };

    this.select = React.createRef();
    this.dropdownRoot = typeof document !== 'undefined' && document.createElement('div');
  }

  componentDidMount() {
    this.props.portal && this.props.portal.appendChild(this.dropdownRoot);
    window.addEventListener('resize', debounce(this.updateSelectBounds));
    window.addEventListener('scroll', debounce(this.onScroll));

    this.props.onChange(this.state.values);

    this.dropDown('close');

    if (this.select) {
      this.updateSelectBounds();
    }
  }

  componentDidUpdate(prevProps, prevState) {
    if (prevState.values !== this.state.values) {
      this.props.onChange(this.state.values);
      this.updateSelectBounds();
    }

    if (prevState.search !== this.state.search) {
      this.updateSelectBounds();
    }

    if (prevState.values !== this.state.values && this.props.closeOnSelect) {
      this.dropDown('close');
    }

    if (prevProps.multi !== this.props.multi) {
      this.updateSelectBounds();
    }

    if (prevState.dropdown && prevState.dropdown !== this.state.dropdown) {
      this.onDropdownClose();
    }

    if (!prevState.dropdown && prevState.dropdown !== this.state.dropdown) {
      this.props.onDropdownOpen();
    }
  }

  componentWillUnmount() {
    this.props.portal && this.props.portal.removeChild(this.dropdownRoot);
    window.removeEventListener(
      'resize',
      debounce(this.updateSelectBounds, this.props.debounceDelay)
    );
    window.removeEventListener('scroll', debounce(this.onScroll, this.props.debounceDelay));
  }

  onDropdownClose = () => {
    this.setState({ cursor: null });
    this.props.onDropdownClose();
  };

  onScroll = () => {
    if (this.props.closeOnScroll) {
      this.dropDown('close');
    }

    this.updateSelectBounds();
  };

  updateSelectBounds = () =>
    this.select.current &&
    this.setState({
      selectBounds: this.select.current.getBoundingClientRect()
    });

  getSelectBounds = () => this.state.selectBounds;

  dropDown = (action = 'toggle', event) => {
    if (
      this.props.portal &&
      !this.props.closeOnScroll &&
      !this.props.closeOnSelect &&
      event &&
      event.srcElement &&
      event.srcElement.offsetParent &&
      event.srcElement.offsetParent.classList.contains('react-dropdown-select-dropdown')
    ) {
      return;
    }

    if (this.props.keepOpen) {
      return this.setState({ dropdown: true });
    }

    if (action === 'close' && this.state.dropdown) {
      this.select.current.blur();

      return this.setState({ dropdown: false, search: this.props.clearOnBlur ? '' : this.state.search });
    }

    if (action === 'open' && !this.state.dropdown) {
      return this.setState({ dropdown: true });
    }

    if (action === 'toggle') {
      this.select.current.focus();
      return this.setState({ dropdown: !this.state.dropdown });
    }

    return false;
  };

  getSelectRef = () => this.select.current;

  addItem = (item) => {
    if (this.props.multi) {
      if (this.state.values.indexOf(item) !== -1) {
        return this.removeItem(null, item, false);
      }

      this.setState({
        values: [...this.state.values, item]
      });
    } else {
      this.setState({
        values: [item],
        dropdown: false
      });
    }

    this.props.clearOnSelect && this.setState({ search: '' });

    return true;
  };

  removeItem = (event, item, close = false) => {
    if (event && close) {
      event.preventDefault();
      event.stopPropagation();
      this.dropDown('close');
    }

    this.setState({
      values: this.state.values.filter(
        (values) => values[this.props.valueField] !== item[this.props.valueField]
      )
    });
  };

  setSearch = (event) => {
    this.setState({
      cursor: null
    });

    this.setState({
      search: event.target.value
    });
  };

  getInputSize = () => {
    if (this.state.search) {
      return this.state.search.length;
    }

    if (this.state.values.length > 0) {
      return this.props.addPlaceholder.length;
    }

    return this.props.placeholder.length;
  };

  toggleSelectAll = () => {
    return this.setState({
      values: this.state.values.length === 0 ? this.selectAll() : this.clearAll()
    });
  };

  clearAll = () => {
    this.props.onClearAll();
    this.setState({
      values: []
    });
  };

  selectAll = () => {
    this.props.onSelectAll();
    return this.setState({
      values: this.props.options.filter((option) => !option.disabled)
    });
  };

  isSelected = (option) =>
    !!this.state.values.find(
      (value) => value[this.props.valueField] === option[this.props.valueField]
    );

  areAllSelected = () =>
    this.state.values.length === this.props.options.filter((option) => !option.disabled).length;

  safeString = (string) => string.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');

  sortBy = (options) => {
    const { sortBy, labelField } = this.props;

    if (!sortBy) {
      return options;
    }

    return options.sort((a, b) => {
      let fieldA;
      let fieldB;

      const sortA = a[sortBy];
      const sortB = b[sortBy];

      if (a[sortBy] === undefined || sortB === undefined) {
        return options;
      }

      if (sortA && typeof sortA === 'number') {
        fieldA = a[sortA ? sortBy : labelField];
        fieldB = b[sortB ? sortBy : labelField];

        return fieldA - fieldB;
      }

      fieldA = String(a[sortA ? sortBy : labelField]).toLowerCase();
      fieldB = String(b[sortB ? sortBy : labelField]).toLowerCase();

      if (fieldA < fieldB) {
        return -1;
      }

      if (fieldA > fieldB) {
        return 1;
      }

      return 0;
    });
  };

  searchResults = () => {
    const regexp = new RegExp(this.safeString(this.state.search), 'i');

    return this.sortBy(this.props.options).filter((item) =>
      regexp.test(item[this.props.searchBy] || item[[this.props.labelField]])
    );
  };

  activeCursorItem = (activeCursorItem) =>
    this.setState({
      activeCursorItem
    });

  handleKeyDown = (event) => {
    const { cursor } = this.state;

    if (event.key === 'ArrowDown' && cursor === null) {
      return this.setState({
        cursor: 0
      });
    }

    if (event.key === 'ArrowUp' || event.key === 'ArrowDown') {
      event.preventDefault();
    }

    if (event.key === 'Escape') {
      this.dropDown('close');
    }

    if (event.key === 'Enter') {
      const currentItem = this.searchResults()[cursor];
      if (currentItem && !currentItem.disabled) {
        this.addItem(currentItem);
      }
    }

    if (event.key === 'ArrowUp' && cursor >= 0) {
      this.setState((prevState) => ({
        cursor: prevState.cursor - 1
      }));
    }

    if (event.key === 'ArrowDown') {
      this.setState((prevState) => ({
        cursor: prevState.cursor + 1
      }));
    }
  };

  renderDropdown = () =>
    this.props.portal ? (
      ReactDOM.createPortal(
        <Dropdown props={this.props} state={this.state} methods={this.methods} />,
        this.dropdownRoot
      )
    ) : (
      <Dropdown props={this.props} state={this.state} methods={this.methods} />
    );

  createNew = (item) => {
    const newValue = {
      [this.props.labelField]: item,
      [this.props.valueField]: item
    };

    this.addItem(newValue);
    this.props.onCreateNew(newValue);
    this.setState({ search: '' });
  };

  render() {
    return (
      <ClickOutHandler onClickOut={(event) => this.dropDown('close', event)}>
        <ReactDropdownSelect
          onKeyDown={this.handleKeyDown}
          onClick={(event) => this.dropDown('open', event)}
          tabIndex="0"
          style={this.props.style}
          ref={this.select}
          disabled={this.props.disabled}
          className={`react-dropdown-select ${this.props.className}`}
          color={this.props.color}>
          <Content props={this.props} state={this.state} methods={this.methods} />

          {this.props.name && <input name={this.props.name} type="hidden" value={this.props.values} />}

          {this.props.loading && <Loading props={this.props} />}

          {this.props.clearable && (
            <Clear props={this.props} state={this.state} methods={this.methods} />
          )}

          {this.props.separator && (
            <Separator props={this.props} state={this.state} methods={this.methods} />
          )}

          {this.props.dropdownHandle && (
            <DropdownHandle
              onClick={() => this.select.current.focus()}
              props={this.props}
              state={this.state}
              methods={this.methods}
            />
          )}

          {this.state.dropdown && this.renderDropdown()}
        </ReactDropdownSelect>
      </ClickOutHandler>
    );
  }
}

Select.defaultProps = {
  addPlaceholder: '',
  placeholder: 'Select...',
  values: [],
  options: [],
  multi: false,
  disabled: false,
  searchBy: 'label',
  sortBy: null,
  clearable: false,
  searchable: true,
  dropdownHandle: true,
  separator: false,
  keepOpen: undefined,
  noDataLabel: 'No data',
  createNewLabel: 'add {search}',
  disabledLabel: 'disabled',
  dropdownGap: 5,
  closeOnScroll: false,
  debounceDelay: 0,
  labelField: 'label',
  valueField: 'value',
  color: '#0074D9',
  keepSelectedInList: true,
  closeOnSelect: false,
  clearOnBlur: true,
  clearOnSelect: true,
  dropdownPosition: 'bottom',
  dropdownHeight: '300px',
  autoFocus: false,
  portal: null,
  create: false,
  name: null,
  onChange: () => undefined,
  onDropdownOpen: () => undefined,
  onDropdownClose: () => undefined,
  onClearAll: () => undefined,
  onSelectAll: () => undefined,
  onCreateNew: () => undefined
};

const ReactDropdownSelect = styled.div`
  position: relative;
  display: flex;
  border: 1px solid #ccc;
  width: 100%;
  border-radius: 2px;
  padding: 2px 5px;
  flex-direction: row;
  align-items: center;
  min-height: 36px;
  ${({ disabled }) =>
    disabled ? 'cursor: not-allowed;pointer-events: none;opacity: 0.3;' : 'pointer-events: all;'}

  :hover, 
  :focus-within {
    border-color: ${({ color }) => color};
  }

  :focus,
  :focus-within {
    outline: 0;
    box-shadow: 0 0 0 3px ${({ color }) => hexToRGBA(color, 0.2)};
  }
`;

export default Select;
