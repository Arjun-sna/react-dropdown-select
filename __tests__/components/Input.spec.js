import React from 'react';
import renderer from 'react-test-renderer';

import Input from '../../src/components/Input';
import { options } from '../../docs/src/options';

const props = {
  parentProps: {
    inputRenderer: null
  },
  parentState: {
    values: options
  },
  parentMethods: {
    getInputSize: () => undefined
  }
};

it('<Input/> renders correctly', () => {
  const tree = renderer.create(<Input {...props}/>).toJSON();

  expect(tree).toMatchSnapshot();
});
