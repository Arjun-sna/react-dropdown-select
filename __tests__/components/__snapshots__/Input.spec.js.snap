// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input/> renders correctly 1`] = `
<input
  className="react-dropdown-select-input css-1mq5oll-InputComponent e11wid6y0"
  onClick={[Function]}
  onKeyPress={[Function]}
  placeholder=""
  readOnly={false}
  tabIndex="-1"
/>
`;
